File: problems/20-valid-parentheses/valid-parentheses.kt
class Solution {
    fun isValid(s: String): Boolean {
        val stack = ArrayDeque<Char>()

        for(char in s){
            when (char) {
                '{', '[', '(' -> stack.addLast(char)
                '}' -> {
                     if (stack.isEmpty()) return false
                    val top = stack.removeLast()
                    if (!(top == '{')){
                        return false
                    }
                }
                ']' -> {
                     if (stack.isEmpty()) return false
                    val top = stack.removeLast()
                    if (!(top == '[')){
                        return false
                    }
                }
                ')' -> {
                     if (stack.isEmpty()) return false
                    val top = stack.removeLast()
                    if (!(top == '(')){
                        return false
                    }
                }
                else -> {
                    return false
                }
                
            }
        }
        return stack.isEmpty()
    }
}

File: problems/20-valid-parentheses/valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        string stack = "";
        for(auto x : s){
            if(x == '(' || x == '[' || x == '{'){
                stack += x;
                continue;
            } 
            if(stack.size() == 0){return false;}
            if(x == ')' && stack[stack.size()-1] == '('){
                stack.pop_back();
                continue;
            } 
            if(x == '}' && stack[stack.size()-1] == '{'){
                stack.pop_back();
                continue;
            } 
            if(x == ']' && stack[stack.size()-1] == '['){
                stack.pop_back();
                continue;
            } 
            return false;
            
        }
        return (stack.size() == 0);
    }
};
